File: WinForms/VB/TreeMapMainDemo/Modules/Grouping.vb
Imports System
Imports System.Data
Imports System.Windows.Forms
Imports DevExpress.Utils

Namespace DevExpress.XtraTreeMap.Demos

    Public Partial Class Grouping
        Inherits TreeMapDemoModuleWithOptions

        Private ReadOnly Property DataAdapter As TreeMapFlatDataAdapter
            Get
                Return CType(treeMapControl1.DataAdapter, TreeMapFlatDataAdapter)
            End Get
        End Property

        Private ReadOnly Property Colorizer As TreeMapPaletteColorizer
            Get
                Return CType(treeMapControl1.Colorizer, TreeMapPaletteColorizer)
            End Get
        End Property

        Public Overrides ReadOnly Property ActiveDemoModule As Control
            Get
                Return treeMapControl1
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
            FillGroupByComboBox()
            DataAdapter.DataSource = CreateDataSet("Billionares.xml")
            Colorizer.Palette = Palette.Office2016Palette
            UpdateGrouping(True)
        End Sub

        Private Sub UpdateGrouping(ByVal enableGrouping As Boolean)
            DataAdapter.GroupDataMembers.Clear()
            If enableGrouping Then
                For Each definition As String In CType(groupByCBE.SelectedItem, GroupDefinitionInfo).Definitions
                    DataAdapter.GroupDataMembers.Add(definition.Replace(" ", ""))
                Next
            End If

            Colorizer.ColorizeGroups = enableGrouping
        End Sub

        Private Sub toolTipController1_BeforeShow(ByVal sender As Object, ByVal e As ToolTipControllerShowEventArgs)
            Dim item As TreeMapItem = CType(e.SelectedObject, TreeMapItem)
            If Not item.IsGroup Then
                Dim superToolTip As SuperToolTip = New SuperToolTip()
                superToolTip.Items.Add(New ToolTipTitleItem() With {.Text = item.Label})
                superToolTip.Items.Add(New ToolTipSeparatorItem())
                Dim dataRowView As DataRowView = CType(item.Tag, DataRowView)
                superToolTip.Items.Add(New ToolTipItem() With {.Text = dataRowView("Residence").ToString() & ", " & dataRowView("NetWorth").ToString() & "B$"})
                e.SuperTip = superToolTip
            End If
        End Sub

        Private Sub FillGroupByComboBox()
            groupByCBE.Properties.Items.AddRange(GroupDefinitionInfo.GetDefinitions())
            groupByCBE.SelectedIndex = 0
        End Sub

        Private Sub groupByCBE_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateGrouping(enableGroupingCE.Checked)
        End Sub

        Private Sub enableGroupingCE_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim enableGrouping As Boolean = enableGroupingCE.Checked
            groupByLCI.Enabled = enableGrouping
            UpdateGrouping(enableGroupingCE.Checked)
        End Sub
    End Class

    Friend Class GroupDefinitionInfo

        Public Shared Function GetDefinitions() As GroupDefinitionInfo()
            Return New GroupDefinitionInfo(2) {New GroupDefinitionInfo(New String() {residenceGroupDefinition}), New GroupDefinitionInfo(New String() {ageGroupDefinition}), New GroupDefinitionInfo(New String() {residenceGroupDefinition, ageGroupDefinition})}
        End Function

        Const residenceGroupDefinition As String = "Residence", ageGroupDefinition As String = "Age Category"

        Private ReadOnly _definitions As String()

        Public ReadOnly Property Definitions As String()
            Get
                Return _definitions
            End Get
        End Property

        Private Sub New(ByVal definitions As String())
            _definitions = definitions
        End Sub

        Public Overrides Function ToString() As String
            Dim definitionsString As String = String.Empty
            For i As Integer = 0 To Definitions.Length - 1
                If Not String.IsNullOrEmpty(definitionsString) Then definitionsString += " And "
                definitionsString += Definitions(i)
            Next

            Return definitionsString
        End Function
    End Class
End Namespace

File: WinForms/CS/TreeMapMainDemo/Modules/Grouping.cs
using System;
using System.Data;
using System.Windows.Forms;
using DevExpress.Utils;

namespace DevExpress.XtraTreeMap.Demos {
    public partial class Grouping : TreeMapDemoModuleWithOptions {

        TreeMapFlatDataAdapter DataAdapter { get { return (TreeMapFlatDataAdapter)treeMapControl1.DataAdapter; } }
        TreeMapPaletteColorizer Colorizer { get { return (TreeMapPaletteColorizer)treeMapControl1.Colorizer; } }
        public override Control ActiveDemoModule { get { return treeMapControl1; } }

        public Grouping() {
            InitializeComponent();
            FillGroupByComboBox();
            DataAdapter.DataSource = Utils.CreateDataSet("Billionares.xml");
            Colorizer.Palette = Palette.Office2016Palette;
            UpdateGrouping(true);
        }

        void UpdateGrouping(bool enableGrouping) {
            DataAdapter.GroupDataMembers.Clear();
            if(enableGrouping)
                foreach(string definition in ((GroupDefinitionInfo)groupByCBE.SelectedItem).Definitions)
                    DataAdapter.GroupDataMembers.Add(definition.Replace(" ", ""));
            Colorizer.ColorizeGroups = enableGrouping;
        }
        void toolTipController1_BeforeShow(object sender, ToolTipControllerShowEventArgs e) {
            TreeMapItem item = (TreeMapItem)e.SelectedObject;
            if(!item.IsGroup) {
                SuperToolTip superToolTip = new SuperToolTip();
                superToolTip.Items.Add(new ToolTipTitleItem() { Text = item.Label });
                superToolTip.Items.Add(new ToolTipSeparatorItem());
                DataRowView dataRowView = (DataRowView)item.Tag;
                superToolTip.Items.Add(new ToolTipItem() { Text = dataRowView["Residence"].ToString() + ", " + dataRowView["NetWorth"].ToString() + "B$" });
                e.SuperTip = superToolTip;
            }
        }
        void FillGroupByComboBox() {
            groupByCBE.Properties.Items.AddRange(GroupDefinitionInfo.GetDefinitions());
            groupByCBE.SelectedIndex = 0;
        }
        void groupByCBE_EditValueChanged(object sender, System.EventArgs e) {
            UpdateGrouping(enableGroupingCE.Checked);
        }
        void enableGroupingCE_CheckedChanged(object sender, EventArgs e) {
            bool enableGrouping = enableGroupingCE.Checked;
            groupByLCI.Enabled = enableGrouping;
            UpdateGrouping(enableGroupingCE.Checked);
        }
    }

    class GroupDefinitionInfo {
        public static GroupDefinitionInfo[] GetDefinitions() {
            return new GroupDefinitionInfo[3] {
                new GroupDefinitionInfo(new string[] { residenceGroupDefinition }),
                new GroupDefinitionInfo(new string[] { ageGroupDefinition }),
                new GroupDefinitionInfo(new string[] { residenceGroupDefinition, ageGroupDefinition })
            };
        }

        const string residenceGroupDefinition = "Residence", ageGroupDefinition = "Age Category";
        readonly string[] _definitions;

        public string[] Definitions { get { return _definitions; } }

        GroupDefinitionInfo(string[] definitions) {
            _definitions = definitions;
        }

        public override string ToString() {
            string definitionsString = string.Empty;
            for(int i = 0; i < Definitions.Length; i++) {
                if(!string.IsNullOrEmpty(definitionsString))
                    definitionsString += " And ";
                definitionsString += Definitions[i];
            }
            return definitionsString;
        }
    }
}
